### DISCUSSION TITLE
Any way to automatically sync deps when running pdm run script?

### DISCUSSION THREAD

**Discussion by Wesztman** [pdm-project/pdm#2667] [Q&A] [answered]
I would like to remove my tox dependency from my project, but it's very simple for new user to just check out and run "tox" only, and the deps are synced automatically.

The new workflow would be to both run "pdm sync" and then "pdm run X" where X is a script with all tests and checks, same as tox runs.

Since tox just use the scripts from pdm/pyproject atm it seems like tox is not really necessary. 

Is there any way to get "pdm run" to sync deps before executing? I.e make the script dependant on a group of deps so to say?

This would mean an equally simple workflow where the user checks out the repo and just run "pdm run X" which then creates venv, install deps and run command. 

It seems that the run command creates a .venv before executing, but there are no deps installed in it.

This would also mean that I would not have multiple venv:s with the same deps since tox creates one and pdm creates one.

Maybe that got confusing... Feel free to ask if I should clarify something :p

Cheers 
Carl

**Comment by frostming** [comment]
```toml
[tool.pdm.scripts]
pre_run = "{pdm} sync"
```

**Reply by Wesztman** [reply]
@frostming 

That works, but it run 3 times times for a composite with 3 scripts, any way to get it to run only one time? 

![image](https://github.com/pdm-project/pdm/assets/54413402/02e46759-a93e-41e1-80e0-784a4a6668a2)

![image](https://github.com/pdm-project/pdm/assets/54413402/076b438a-92a3-4a5b-a414-7446ec4dc5ce)

**EDIT:** Changed to "pre_all" that works a little better 😊

![image](https://github.com/pdm-project/pdm/assets/54413402/90b7546c-a7f3-4090-acac-533cd0c63a16)

 But there is still alot of clutter, any way to not show the info prints? Tried to find a log level setting with no success 🤔
 
![image](https://github.com/pdm-project/pdm/assets/54413402/ed31aea1-3f0f-43ef-862c-47260abde672)

While I'm at it, a question on the `INFO: The saved Python interpreter doesn't match the project's requirement. Trying to find another one.` . I'm using Python 3.12.2 and my pyproject.toml sais `requires-python = ">=3.9"`. Am I missing something if I get this print? 🤔 Because the python used should satisfy the requirements.

**Reply by frostming** [reply, marked as answer]
It runs on every invocation of `pdm run`, actually you don't need to call `pdm run` in tasks:

```toml
[tool.pdm.scripts]
pre_run = "{pdm} sync"
ruff-lint = "ruff check . --fix"
mypycheck = "mypy ."
lint = { composite = ["ruff-lint", "mypycheck"] }
```

Your solution `pre_all` only runs when running task `all`

**Reply by frostming** [reply]
> While I'm at it, a question on the `INFO: The saved Python interpreter doesn't match the project's requirement. Trying to find another one.`

You may have saved an incorrect python path in `.pdm-python`, check and remove it if necessary.

**Reply by Wesztman** [reply]
@frostming 

Thanks for all the answers :) last question, is there any configuration to not output the info logging prints?

**Reply by Wesztman** [reply]
> > While I'm at it, a question on the `INFO: The saved Python interpreter doesn't match the project's requirement. Trying to find another one.`
> 
> You may have saved an incorrect python path in `.pdm-python`, check and remove it if necessary.

The problem was that in my devcontainer I had set 

![image](https://github.com/pdm-project/pdm/assets/54413402/e9981e90-b207-4eb1-88f1-28fd58e4ebba)

But ${workspaceFolder} apparently does not expand properly in containerEnv for some reason.

![image](https://github.com/pdm-project/pdm/assets/54413402/113e3499-f962-465e-879d-7250f2b85d58)

Removing the PDM_PYTHON env var solved the issue (I don't need it anymore anyway).

**Reply by Wesztman** [reply]
Now it works like a charm 😄 final code 

![image](https://github.com/pdm-project/pdm/assets/54413402/670c41b0-528f-4f67-b4be-f58602796ec2)

Works like a charm 

![image](https://github.com/pdm-project/pdm/assets/54413402/ad647f8c-45fb-44f8-b1ad-e8759bc05925)

I might do a PR sometime to add a "title" option to scripts, something like 

```
[tool.pdm.scripts]
typecheck.cmd = "mypy ."
typecheck.title = "Running type checks"
```

Do you think it would be a feature that would get merged? 😊

**Reply by frostming** [reply]
> last question, is there any configuration to not output the info logging prints?


`--quiet`

> I might do a PR sometime to add a "title" option to scripts, something like

What is it for?    We have `help` if  that is what you need

**Reply by Wesztman** [reply]
> What is it for? We have `help` if that is what you need

It would just add a heading/title line to the script output as I've done in my code, without needing an extra script "add-title" like I have 😊

Adding a clear boundary between the scripts and its outputs when running composites.

**Reply by frostming** [reply]
That's too verbose, users are free to add any logging to the console in their script. And we display the current running commands if you provide `-v` flag

**Reply by Wesztman** [reply]
> That's too verbose, users are free to add any logging to the console in their script. And we display the current running commands if you provide `-v` flag

Sounds reasonable 😊 thanks for all the help!
